message_sending: Show retry/cancel buttons only after message fails

x15sr71 opened the pull request:
**Fixes:** Improve part 1 of the issue #29100  

**Description:**  
This PR addresses the remaining issue from #29100 by adding a condition to ensure that the Retry and Cancel buttons only appear when a message has truly failed. The buttons will no longer be shown while the message is still sending (i.e., awaiting server response), as they would not function correctly in that state.


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


<details>
<summary>Self-review checklist</summary>

- [x] [[[Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code)](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code)](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability (variable names, code reuse, readability, etc.).
  Communicate decisions, questions, and potential concerns.
  
- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

- [ ] Individual commits are ready for review (see [[[commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).
  - [ ] Each commit is a coherent idea.
  - [ ] Commit message(s) explain reasoning and motivation for changes.

- [ ] Completed manual review and testing of the following:
  - [ ] Visual appearance of the changes.
  - [ ] Responsiveness and internationalization.
  - [ ] Strings and tooltips.
  - [ ] End-to-end functionality of buttons, interactions, and flows.
  - [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

berelian1 commented:
@zulipbot claim

alya commented:
Thanks for the fix! Could you please update your commit message to match the [commit style guidelines](https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-messages)? 

To save reviewers time, please keep your PR description to the point. Have you had a chance to explore GitHub's UI yet? It's easy to see what code changes a PR makes.

x15sr71 commented:
Thanks @alya for the feedback! I’ve updated the commit message, PR title, and trimmed the description; please let me know if there’s anything else to adjust.

x15sr71 commented:
Thanks, @alya, for the feedback! I’ve updated the commit message, PR title, and trimmed the description. I’ll be more mindful of the community guidelines moving forward. Please let me know if there’s anything else to adjust

alya commented:
The commit message definitely needs clarification now. Also, the commit message and PR description should indicate that this PR fixes #29100, as it's intended to fix the only remaining open point.

amanagr commented:
@x15sr71 can you also add screenshot  / recording which shows how you tested this.

x15sr71 commented:
Hi @amanagr, I wasn’t able to reproduce the first part of the [issue](https://github.com/zulip/zulip/issues/29100#issue-2157680466), as I noted in detail in my [comment](https://github.com/zulip/zulip/issues/29100#issuecomment-2764230796). However, after extensive code reading and thought, I felt it was still worthwhile to add a safeguard, so the changes in this PR handle rare edge cases where the problem could occur. Let me know if anything else is needed!

x15sr71 commented:
Hi @amanagr, just checking in to see if you’ve had a chance to review this PR. I’ve incorporated the requested changes and added the clarification in the comments as well. Please let me know if there’s anything else I should address. Thank you!

amanagr commented on web/src/echo.ts:
```diff
@@ -125,11 +125,16 @@ function hide_retry_spinner($row: JQuery): boolean {
         $row.find(".slow-send-spinner").addClass("hidden");
-        const $message_controls = $row.find(".message_controls");
-        $message_controls.html(render_message_controls_failed_msg());
+        $row.find(".message_controls").html(render_message_controls_failed_msg());
```
I don't understand the reason for this change.

x15sr71 replied:
I replaced the variable assignment with a direct update to $row.find(".message_controls").html(render_message_controls_failed_msg()) to optimize the code by eliminating unnecessary variables. However, I now understand that this change wasn’t necessary.

amanagr commented on web/src/echo.ts:
```diff
@@ -125,11 +125,16 @@ function hide_retry_spinner($row: JQuery): boolean {
-    // Failed to send message, so display inline retry/cancel
+    // Don't show retry/cancel buttons if the message doesn't exist or is still
+    // in the sending state (locally_echoed but not yet failed)
+    const message = message_store.get(message_id);
+    if (!message || (message.locally_echoed && !message.failed_request)) {
+        return;
```
We don't need this condition since the caller already ensure these conditions.

amanagr commented:
Closing since the PR doesn't address the issue.

x15sr71 commented:
Thank you for your feedback @amanagr on the PR. Since it has been closed, I’ll proceed with the next issue. I wasn’t able to reproduce the issue, which is why I added the safeguard in show_message_failed. I’ve been considering the possibility that the issue may already have been resolved or may not exist. I’m always eager to learn from these situations, so if you have any additional context or steps to reproduce, I’d really appreciate it if you could share them for future reference.

x15sr71 commented:
Thank you for your feedback @amanagr on the PR. Since it has been closed, I’ll proceed with the next issue. I wasn’t able to reproduce the issue, which is why I added the safeguard in show_message_failed. I’ve been considering the possibility that the issue may already have been resolved or may not exist. I’m always eager to learn from these situations, so if you have any additional context or steps to reproduce, I’d really appreciate it if you could share them for future reference.

Thank you once again.
